nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/keyvault/AzureKeyVaultClientService.java

```diff
@@ -0,0 +1,242 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.services.azure.keyvault;
+
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.ValidationContext;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.util.StringUtils;
+
+import com.azure.identity.DefaultAzureCredentialBuilder;
+import com.azure.identity.ClientSecretCredential;
+import com.azure.identity.ClientSecretCredentialBuilder;
+import com.azure.security.keyvault.secrets.SecretClient;
+import com.azure.security.keyvault.secrets.SecretClientBuilder;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@Tags({"azure", "keyvault", "credential", "service", "secure"})
+@CapabilityDescription(
+        "Provides a controller service that configures a connection to Azure Key Vault" +
+                " and provides access to that connection to Azure Key Vault components."
+)
+public class AzureKeyVaultClientService
+        extends AbstractControllerService
+        implements AzureKeyVaultConnectionService {
+
+    private String keyVaultName;
+    private String servicePrincipalClientID;
+    private String servicePrincipalClientSecret;
+    private String tenantID;
+    private String endPointSuffix;
+    private Boolean useManagedIdentity;
+    private SecretClient keyVaultSecretClient;
+    private ComponentLog logger;
+    private LoadingCache<String, String> secretCache;
+
+    public String getKeyVaultName() {
+        return this.keyVaultName;
+    }
+
+    public String getServicePrincipalClientID() {
+        return this.servicePrincipalClientID;
+    }
+
+    public String getServicePrincipalClientSecret() {
+        return this.servicePrincipalClientSecret;
+    }
+
+    public String getTenantID() {
+        return this.tenantID;
+    }
+
+    public String getEndPointSuffix() {
+        return this.endPointSuffix;
+    }
+
+    public Boolean getUseManagedIdentity() {
```

Comment:
Are the above public get methods necessary?  In general, public methods on Controller Service implementations that do not implement interface methods should be avoided.

Reply:
cleaned